hibernate-core/src/main/java/org/hibernate/dialect/OracleOsonJacksonArrayJdbcType.java

```diff
@@ -0,0 +1,178 @@
+			osonFactory = osonFactoryKlass.getDeclaredConstructor().newInstance();
+		}
+		catch (Exception | LinkageError e) {
+			// should not happen as OracleOsonJacksonJdbcType is loaded
+			// only when Oracle OSON JDBC extension is present
+			// see OracleDialect class.
+			throw new ExceptionInInitializerError( "OracleOsonJacksonJdbcType class loaded without OSON extension: " + e.getClass()+" "+ e.getMessage());
+		}
+	}
+
+
+	public OracleOsonJacksonArrayJdbcType(JdbcType elementJdbcType) {
+		super(elementJdbcType);
+	}
+
+
+	@Override
+	public String toString() {
+		return "OracleOsonJacksonArrayJdbcType";
+	}
+
+
+	@Override
+	public <X> ValueBinder<X> getBinder(JavaType<X> javaType) {
+
+		return new BasicBinder<>( javaType, this ) {
+
+			private <X> byte[] toOsonStream(X value, JavaType<X> javaType, WrapperOptions options) throws Exception {
+				final Object[] domainObjects = javaType.unwrap( value, Object[].class, options );
+
+				ByteArrayOutputStream out = new ByteArrayOutputStream();
+				try (OracleJsonGenerator generator = new OracleJsonFactory().createJsonBinaryGenerator( out )) {
+					OsonDocumentWriter writer = new OsonDocumentWriter( generator );
+
+					if ( getElementJdbcType() instanceof JsonJdbcType jsonElementJdbcType ) {
+						final EmbeddableMappingType embeddableMappingType = jsonElementJdbcType.getEmbeddableMappingType();
+						JsonHelper.serializeArray( embeddableMappingType, domainObjects, options, writer );
+					}
+					else {
+						assert !( getElementJdbcType() instanceof AggregateJdbcType );
+						final JavaType<?> elementJavaType = ( (BasicPluralJavaType<?>) javaType ).getElementJavaType();
+						JsonHelper.serializeArray(
+								elementJavaType,
+								getElementJdbcType(),
+								domainObjects,
+								options,
+								writer
+						);
+					}
+					generator.close();
+					return out.toByteArray();
+				}
+
+			}
+			@Override
+			protected void doBind(PreparedStatement st, X value, int index, WrapperOptions options)
+					throws SQLException {
+				try {
+					st.setObject( index, toOsonStream( value, getJavaType(), options ), OracleType.JSON );
+				}
+				catch (Exception e) {
+					throw new SQLException( e );
+				}
+			}
+
+			@Override
+			protected void doBind(CallableStatement st, X value, String name, WrapperOptions options)
+					throws SQLException {
+				try {
+					st.setObject( name, toOsonStream( value, getJavaType(), options ) , OracleType.JSON);
+				}
+				catch (Exception e) {
+					throw new SQLException( e );
+				}
+			}
+		};
+	}
+
+	@Override
+	public <X> ValueExtractor<X> getExtractor(JavaType<X> javaType) {
+
+		return new BasicExtractor<>( javaType, this ) {
+
+			private X fromOson(InputStream osonBytes, WrapperOptions options) throws Exception {
+				FormatMapper mapper = options.getJsonFormatMapper();
+				JsonParser osonParser = ((JsonFactory)osonFactory).createParser( osonBytes );
+				return mapper.readFromSource(  getJavaType(), osonParser, options);
```

Comment:
Looks like your missing something here. This should delegate to a `JsonHelper` method. Please see this comment which was hinting at how you could extract parts of the `JsonHelper.arrayFromString` method which you can call here: https://github.com/hibernate/hibernate-orm/pull/9629#discussion_r1950936864